cirq/sim/simulator.py

```diff
@@ -457,11 +457,13 @@ def simulate_moment_steps(
             Iterator that steps through the simulation, simulating each
             moment and returning a StepResult for each moment.
         """
-        return self._simulator_iterator(
-            circuit, study.ParamResolver(param_resolver), qubit_order, initial_state
-        )
+        param_resolver = study.ParamResolver(param_resolver)
+        resolved_circuit = protocols.resolve_parameters(circuit, param_resolver)
+        check_all_resolved(resolved_circuit)
+        actual_initial_state = 0 if initial_state is None else initial_state
+        return self._base_iterator(resolved_circuit, qubit_order, actual_initial_state)
```

Comment:
@balopat  This ends up being a breaking change though for any simulator that was overriding `_simulator_iterator`. This is why I had left the call to `_simulator_iterator` there in the original PR #3650. Otherwise I'd have just inlined and deleted `_simulator_iterator` outright: it's a private method and it's unused elsewhere.

It doesn't sound like anyone has been broken by this, so maybe it's okay. (We can just delete `_simulator_iterator` in that case; it's not called by anyone). I think there's still an unmet need for cleanly deprecate abstract methods, or methods that were intended to be overridden. The base class has to continue calling them until the deadline or it will break child classes. I like the error-by-default, but maybe there's a way to silence these errors when retaining the call to a deprecated method is required, like here.

Reply:
You are totally right, I rushed this one this doesn't make sense at all as it is :).
I'll add the BREAKING_CHANGE label to this PR. Also, I'll add it to the list of breaking changes in the release notes for completeness sake. 

Re abstract methods deprecation - hmm yeah we'll need to think about that. I'll open an issue for it.